### ISSUE TITLE
Retrieving gradients in different threads panics.

### ISSUE THREAD

**Issue by giucesar** [tracel-ai/burn#1142]
Hi.

If a tensor is created in the main thread, its gradients panic when the operations happen in a different thread.

Example:
```
use std::{thread::sleep, time::Duration};

use burn::{backend::Autodiff, tensor::Tensor};

type B = Autodiff<burn::backend::NdArray>;

fn obj(x: Tensor<B, 1>) -> Tensor<B, 1> {
    (x.clone() * x.clone()).sum().sqrt()
}

fn main() {
    let x = Tensor::<B, 1>::from_floats([1.0, 2.0, 3.0]).set_require_grad(true);
    let op = move || {
        let value = obj(x.clone()).sum();
        let grads = value.backward();
        let g_x = Tensor::<B, 1>::from_inner(x.clone().grad(&grads).unwrap());

        let x_1 = x.clone() - g_x;
        let value = obj(x_1.clone()).sum();
        println!("value: {}", value.into_scalar());
    };

    op.clone()();

    let thread_1 = std::thread::spawn(op.clone());
    // sleep(Duration::from_millis(100));
    let thread_2 = std::thread::spawn(op.clone());
    thread_1.join().unwrap();
    thread_2.join().unwrap();
}
```

If you run this small code several times, it fails most of the time in the line that retrieves the gradients of x. 
The error message is:
```
Root node should have a step registered, did you forget to call `Tensor::register_grad` on the tensor where you need gradients?
```

Interestingly, if you comment the thread_2, it works as expected. Also, inserting 
```
sleep(Duration::from_millis(100));
```
between the thread spawns makes it work as expected.

Here is the full backtrace:
```
value: 2.7416575
thread '<unnamed>' panicked at src/main.rs:16:69:
called `Option::unwrap()` on a `None` value
value: 2.7416575
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:127:5
   3: core::option::Option<T>::unwrap
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/option.rs:931:21
   4: test_burn::main::{{closure}}
             at ./src/main.rs:16:46
   5: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
thread 'main' panicked at src/main.rs:28:21:
called `Result::unwrap()` on an `Err` value: Any { .. }
stack backtrace:
   0: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   2: core::result::unwrap_failed
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1077:23
   4: test_burn::main
             at ./src/main.rs:28:5
   5: core::ops::function::FnOnce::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by giucesar** [comment]
To add.
Searching "register_grad" in the source code only returns the string defining this error message.

**Comment by nathanielsimard** [comment]
Reviewing the provided code, it becomes evident that the failure is expected with the current implementation. The issue lies in the fact that the root graph node is created only once (before the op is generated). Consequently, there is a single graph at that point. However, when you execute op twice, the graph is not splitted into two distinct entities, both operations share the same graph and modify it. It's important to note that you can only invoke the backward operation once on a graph, as the graph gets consumed in the process. While the same graph can be safely modified by multiple threads, it should not be consumed by both threads simultaneously.

The solution to this problem is to ensure that each thread has its own graph. This can easily be done with the `detach` method.

```rust
use std::{thread::sleep, time::Duration};

use burn::{backend::Autodiff, tensor::Tensor};

type B = Autodiff<burn::backend::NdArray>;

fn obj(x: Tensor<B, 1>) -> Tensor<B, 1> {
    (x.clone() * x.clone()).sum().sqrt()
}

fn main() {
    let x = Tensor::<B, 1>::from_floats([1.0, 2.0, 3.0]).set_require_grad(true);
    let op = move || {
        // detach will create a new graph here.
        let x = x.detach();
        // Same as before.
        let value = obj(x.clone()).sum();
        let grads = value.backward();
        let g_x = Tensor::<B, 1>::from_inner(x.clone().grad(&grads).unwrap());

        let x_1 = x.clone() - g_x;
        let value = obj(x_1.clone()).sum();
        println!("value: {}", value.into_scalar());
    };

    op.clone()();

    let thread_1 = std::thread::spawn(op.clone());
    // sleep(Duration::from_millis(100));
    let thread_2 = std::thread::spawn(op.clone());
    thread_1.join().unwrap();
    thread_2.join().unwrap();
}
```

Note that sometimes your code _may_ work as intended because the creation of the graph is actually lazy. So, when a graph doesn't exist, we create a new one. This is why you can call your code multiple times sequentially, and it will work.

**Comment by giucesar** [comment]
It makes sense, thanks.
That is pretty hard to identify without knowing the internals of the library. Honestly, I had an incorrect understanding of the detach operation.

**Comment by nathanielsimard** [comment]
To be fair, I believe we should enhance the error message and generate additional documentation on how to utilize multithreading with autodiff. Subsequently, we could provide a link to the documentation within the error message, thereby enhancing its usefulness.

**Comment by nathanielsimard** [comment]
You are updating each tensor in the list; however, the newly updated tensors aren't tagged as `required grad`. Therefore, in the second iteration, you need to call `require_grad`. Keep in mind that you aren't _just_ updating the tensor value with the mutation, but all the tensor state that comes with it. It's like creating new parameters at each new iteration; this is why a new `require_grad` is necessary.

**Comment by giucesar** [comment]
Understood. I incorrectly assumed it kept the require_grad tag. My code was previously written with pure tch, and it had the += operator, which keeps the gradient tag.
